devlib/instrument/schedstats.py

```diff
@@ -0,0 +1,166 @@
+    schedstat_path = '/proc/schedstat'
+
+    def setup(self):
+        try:
+            self.old_sysctl_value = self.target.read_int(self.sysctl_path)
+        except TargetError:
+            if not self.target.file_exists(self.sysctl_path):
+                raise TargetError('schedstats not supported by target. '
+                                  'Ensure CONFIG_SCHEDSTATS is enabled.')
+
+        self.target.write_value(self.sysctl_path, 1)
+
+        # Check version matches
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        match = re.search(r'version ([0-9]+)', lines[0])
+        if not match or match.group(1) != '15':
+            raise TargetError(
+                'Unsupported schedstat version string: "{}"'.format(lines[0]))
+
+        # Take a sample of the schedstat file to figure out which channels to
+        # create.
+        # We'll create a site for each CPU and a site for each sched_domain.
+        for site, measures in self._get_sample().iteritems():
+            if site.startswith('cpu'):
+                measurement_category = 'schedstat_cpu'
+            else:
+                measurement_category = 'schedstat_domain'
+
+            for measurement_name in measures.keys():
+                # TODO: constructing multiple MeasurementTypes with same
+                # params. Does it matter?
+                measurement_type = MeasurementType(
+                    measurement_name, '', measurement_category)
+                self.add_channel(site=site,
+                                 name='{}_{}'.format(site, measurement_name),
+                                 measure=measurement_type)
+
+    def teardown(self):
+        self.target.write_value(self.sysctl_path, self.old_sysctl_value)
+
+    def _get_sample(self):
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        ret = OrderedDict()
+
+        # Example /proc/schedstat contents:
+        #
+        # version 15
+        # timestamp <timestamp>
+        # cpu0 <cpu fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+        # cpu1 <cpu_fields>
+        # domain0 <domain fields>
+        # domain1 <domain fields>
+
+        curr_cpu = None
+        for line in lines[2:]:
+            tokens = line.split()
+            if tokens[0].startswith('cpu'):
+                curr_cpu = tokens[0]
+                site = curr_cpu
+                measures = CPU_MEASURES
+            elif tokens[0].startswith('domain'):
+                if not curr_cpu:
+                    raise TargetError(
+                        'Failed to parse schedstats, found domain before CPU')
+                # We'll name the site for the domain like "cpu0domain0"
+                site = curr_cpu + tokens[0]
+                measures = DOMAIN_MEASURES
+            else:
+                raise TargetError(
+                    'Unrecognised schedstats line: "{}"'.format(line))
```

Comment:
I started doing part 2. and realised that it's also pretty unpalatable; the parsing state (`curr_cpu` and `curr_domain`) makes the code quite arcane and it still isn't generic. If we want to keep fiddling with schedstats we're going to have to fiddle with this code again anyway so I propose just skipping to part 3.